buildsystem/generator/make.go

```diff
@@ -0,0 +1,300 @@
+package generator
+
+import (
+	"io/ioutil"
+
+	"github.com/evergreen-ci/shrub"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/grip"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+// Make represents an evergreen config generator for Makefile-based projects.
+type Make struct {
+	Tasks    map[string]MakeTask    `yaml:"tasks,omitempty"`
+	Variants map[string]MakeVariant `yaml:"variants"`
+	// Environment defines global environment variables. Definitions can be
+	// overridden at the task or variant level.
+	Environment      map[string]string `yaml:"environment,omitempty"`
+	WorkingDirectory string            `yaml:"-"`
+}
+
+// NewMake creates a new evergreen config generator for Make from a single file
+// that contains all the necessary generation information.
+func NewMake(file string, workingDir string) (*Make, error) {
```

Comment:
is it time to bring back the interface. I'm iffy about having the *Make type exported.

Reply:
We could interface this but the only commonality would be the `Generate` function. There's almost no shared functionality between building an evergreen config from a Make generator and building an evergreen config from a Golang generator except their output. The actual implementation needs of the different generators have almost no similarity because they need different definition files to be merged in.

I would prefer to just duck-type it and if someone wants to say that the only functionality it has is `Generate()`, then they can define their own interface that gates the Make/Golang generator functionality. I don't actually like to return interfaces when you could just receive the concrete type and compile-time check that it satisfies whatever interface you want to put it in.

Also, returning a limited interface would make the difficult to use unless we made a separate model for the YAML file representation (with no interface), marshalled that representation into raw bytes, and then passed those raw bytes into the Make generator to generate evergreen config.